### ISSUE TITLE
Doubt regarding link.txt produced using target_link_libraries

### ISSUE THREAD

**Issue by anutosh491** [emscripten-core/emscripten#22811]
I am working towards producing a static library (`libclangCppInterOp.a`). In my CMakeLists.txt file I have this (the libraries `clanginterpeter` etc are obtained by building llvm( clang, lld) for wasm)
```
set(LLVM_LINK_COMPONENTS
  ${LLVM_TARGETS_TO_BUILD}
  BinaryFormat
  Core
  Object
  OrcJit
  Support
  MC
  TargetParser
  WebAssembly
)

set(link_libs
  clanginterpreter
  clangAST
  clangBasic
  clangFrontend
  clangLex
  clangSema
  )
  
  add_llvm_library(clangCppInterOp
  DISABLE_LLVM_LINK_LLVM_DYLIB
  CppInterOp.cpp
  ${DLM}
  LINK_LIBS
  ${link_libs}
 )
 ```
 Now under the hood add_llvm_library does this (shown by `--trace-expand`)
 ```
 add_library(clangCppInterOp STATIC CppInterOp.cpp;DynamicLibraryManager.cpp;DynamicLibraryManagerSymbol.cpp;Paths.cpp;/Users/anutosh491/work/recipes/output/bld/rattler-
 │ │ build_cppinterop_1730189487/work/lib/Interpreter/Compatibility.h;/Users/anutosh491/work/recipes/output/bld/rattler-build_cppinterop_1730189487/work/lib/Interpreter/CppInterOpInterpreter.h;/Users/anutosh491/work/r
 │ │ ecipes/output/bld/rattler-build_cppinterop_1730189487/work/lib/Interpreter/DynamicLibraryManager.h;/Users/anutosh491/work/recipes/output/bld/rattler-build_cppinterop_1730189487/work/lib/Interpreter/Paths.h )
 
 target_link_libraries(clangCppInterOp PUBLIC clangInterpreter;clangAST;clangBasic;clangFrontend;clangLex;clangSema;lldWasm;LLVMWebAssemblyCodeGen;LLVMWebAssemblyAsmPars
 │ │ er;LLVMWebAssemblyDesc;LLVMWebAssemblyDisassembler;LLVMWebAssemblyInfo;LLVMWebAssemblyUtils;LLVMBinaryFormat;LLVMCore;LLVMObject;LLVMOrcJIT;LLVMSupport;LLVMMC;LLVMTargetParser;LLVMWebAssemblyCodeGen;LLVMWebAssemb
 │ │ lyAsmParser;LLVMWebAssemblyDesc;LLVMWebAssemblyDisassembler;LLVMWebAssemblyInfo;LLVMWebAssemblyUtils;LLVMFrontendDriver;LLVMOrcDebugging;dl  LLVMWebAssemblyCodeGen;LLVMWebAssemblyAsmParser;LLVMWebAssemblyDesc;LLV
 │ │ MWebAssemblyDisassembler;LLVMWebAssemblyInfo;LLVMWebAssemblyUtils;LLVMBinaryFormat;LLVMCore;LLVMObject;LLVMOrcJIT;LLVMSupport;LLVMMC;LLVMTargetParser;LLVMWebAssemblyCodeGen;LLVMWebAssemblyAsmParser;LLVMWebAssembl
 │ │ yDesc;LLVMWebAssemblyDisassembler;LLVMWebAssemblyInfo;LLVMWebAssemblyUtils;LLVMFrontendDriver;LLVMOrcDebugging )

 ```
 Which looks like everything I want 
 1) First add_library with all .cpp.o files
 2) link against certain libraries
 
 But when I see the link.txt file I see this being generated
 ```
opt/emsdk/upstream/emscripten/emar qc libclangCppInterOp.a CMakeFiles/clangCppInterOp.dir/CppInterOp.cpp.o CMakeFiles/clangCppInterOp.dir/DynamicLibraryManager.cpp.o CMakeFiles/clangCppInterOp.dir/DynamicLibraryManagerSymbol.cpp.o CMakeFiles/clangCppInterOp.dir/Paths.cpp.o
opt/emsdk/upstream/emscripten/emranlib libclangCppInterOp.a
```

Which means I lost step 2 I think (have I ?)

So technically the file of interest being [this](https://github.com/compiler-research/CppInterOp/blob/main/lib/Interpreter/CMakeLists.txt) , removing everything unnecessary would leave us with this
```
add_llvm_library(clangCppInterOp
  CppInterOp.cpp
  DynamicLibraryManager.cpp
  DynamicLibraryManagerSymbol.cpp
  Paths.cpp

  # Specify LLVM components directly for linking
  LINK_COMPONENTS
  BinaryFormat
  Core
  Object
  OrcJit
  Support
  MC
  TargetParser
  WebAssembly

  # Additional libraries from Clang and LLD
  LINK_LIBS
  clangInterpreter
  clangAST
  clangBasic
  clangFrontend
  clangLex
  clangSema
  lldWasm
)
```
Still I don't think we end up having `clanginterpreter` etc in the `link.txt` file.
So what I fail to understand is that even if the `link.txt` is correctly formed ... where should the linking between `libClangCppInterop` and `libClangInterpreter` take place ? Am I missing something that doesn't conduct that linkage ?

Basically I should be able to verify the linking through something like this (currently fails)
```
(xeus-cpp-wasm-host) anutosh491@Anutoshs-MacBook-Air lib % emnm libclangInterpreter.a | grep '__clang_Interpreter_SetValueNoAlloc' 
00005cf4 T __clang_Interpreter_SetValueNoAlloc
(xeus-cpp-wasm-host) anutosh491@Anutoshs-MacBook-Air lib % emnm libclangCppInterOp.a | grep '__clang_Interpreter_SetValueNoAlloc'. 
```

**Comment by anutosh491** [comment]
Am i missing something here? I tried not using `add_llvm_library` provided by llvm and simply did this 
```
add_library(clangCppInterOp STATIC 
 CppInterOp.cpp 
 DynamicLibraryManager.cpp 
 DynamicLibraryManagerSymbol.cpp 
 Paths.cpp
)

target_link_libraries(clangCppInterOp PUBLIC
 clangInterpreter 
 clangAST 
 clangBasic 
 clangFrontend 
 clangLex 
 clangSema 
 lldWasm
 embind
)

target_link_options(clangCppInterOp PRIVATE
 -O1
 -sMODULARIZE
 -sEXPORT_ES6=1
 -sASSERTIONS
 -sALLOW_MEMORY_GROWTH=1
 -sINITIAL_MEMORY=128MB
 -sTOTAL_STACK=32MB
 -sMAIN_MODULE=1
 -sEXPORTED_RUNTIME_METHODS=ccall,cwrap,stringToNewUTF8,getValue,setValue
 -sEXPORTED_FUNCTIONS=_malloc,_free,__ZTIN10emscripten3valE
 --preload-file ${EMSCRIPTEN_SYSROOT}/@/
)
```

But that didn't help me either. I still lose any information related to the libraries and the flags.

**Comment by anutosh491** [comment]
While not building against emscripten (as in for a non-wasm build) the correct link.txt is generated (though CppInterOp does generate a shared object) ... something like this 
```
/usr/bin/g++-12 -fPIC  ..... lot of flags ..... -o ../libclangCppInterOp.so.18.1 CMakeFiles/clangCppInterOp.dir/CppInterOp.cpp.o  .. other object files... CMakeFiles/clangCppInterOp.dir/Paths.cpp.o  -Wl,-rpath,"\$ORIGIN/../lib:/home/runner/work/CppInterOp/CppInterOp/llvm-project/build/./lib:" /home/runner/work/CppInterOp/CppInterOp/llvm-project/build/lib/libclangInterpreter.a ... followed by other libraries
```
So add_llvm_library only doesn't work out while building against emscripten is what I can see.

**Comment by anutosh491** [comment]
Opened something relevant here too (https://github.com/llvm/llvm-project/issues/114223)

**Comment by anutosh491** [comment]
@sbc100 @kripken could y'all maybe help me out here ?

**Comment by sbc100** [comment]
I think maybe the confusion related to differences between static libraries (built with ar/emar) and dynamic libraries (built with `em++ -shared`/`g++ -shared`).    

Regardless of which compiler you use static libraries cannot express dependencies on each other.  Its not generally possible to link one static library again another or embed one in another.   Instead those dependencies all have to be propagated to the final binary.  So if you are trying to link against libfoo.a, and libfoo.a happens to to internally depend on libbar.a you need to add both `-lfoo` and `-lbar` to your final link command.

Shared libraries on the other hand can have transitive dependencies and so you would only need to link with `-lfoo` when using dynamic linking.

With emscripten, using static linking is much more efficient and simpler and highly recommended (so you are doing the right thing by building statically using `emar` in this case).

**Comment by anutosh491** [comment]
Yes that's very much related to my doubt. So just to confirm I understand correctly .... I have libA.a that should only be dependent on libB.a (and possibly be directly dependent on libC.a) in that case I can't propagate dependencies as I've written in the issue ? Is that correct?

So basically if I want to A to be completely independent on C I have to use LibB.so instead of LibB.a ?

**Comment by sbc100** [comment]
> Yes that's very much related to my doubt. So just to confirm I understand correctly .... I have libA.a that should only be dependent on libB.a (and possibly not be directly dependent on libC.a) in that case I can't propagate dependencies as I've written in the issue (basically propagate C through B) ? Is that correct?
> 
> So basically if I want to A to be completely independent on C I have to use LibB.so instead of LibB.a ?

The way it normally works is that with static linking you just need to explicitly list of the dependencies when linking you final executable.

So if A depends on B and B depends on C then you need go `gcc libA.a libB.a libC.a -o myprog` and doing just `gcc libA.a -o myprog` will not work.

I'm sure llvm can be built statically so there must be some way that add_llvm_library takes care of this internally and automatically does this transitive dependency stuff.

BTW, none of this is specific to emscripten, this is just the way that static linking works in gcc / clang / etc.

**Comment by anutosh491** [comment]
> I'm sure llvm can be built statically so there must be some way that add_llvm_library takes care of this internally and automatically does this transitive dependency stuff.
>
> BTW, none of this is specific to emscripten, this is just the way that static linking works in gcc / clang / etc.

Yes this is exactly where are my doubts are coming from.

So I am maintaining the recipe for llvm on emscripten-forge ([build.sh](https://github.com/emscripten-forge/recipes/blob/main/recipes/recipes_emscripten/llvm/build.sh)). Now llvm obviously has lot of `add_llvm_library`, `add_clang_library` calls and lot of static libraries depending on other static libraries.
So let's take a static library like (libclangInterpreter.a)

I see the following 
```
add_clang_library(clangInterpreter
  // source files like Interpreter.cpp

  DEPENDS
  intrinsics_gen
  ClangDriverOptions

  LINK_LIBS
  // lot of static libraries like clangAST

  )
  ```
  
  Now this leads to the following in link.txt
  ```
│ │ $BUILD_PREFIX/opt/emsdk/upstream/emscripten/emar qc ../../../../lib/libclangInterpreter.a CMakeFiles/obj.clangInterpreter.dir/DeviceOffload.cpp.o CMakeFile
 │ │ s/obj.clangInterpreter.dir/CodeCompletion.cpp.o CMakeFiles/obj.clangInterpreter.dir/IncrementalExecutor.cpp.o CMakeFiles/obj.clangInterpreter.dir/Increment
 │ │ alParser.cpp.o CMakeFiles/obj.clangInterpreter.dir/Interpreter.cpp.o CMakeFiles/obj.clangInterpreter.dir/InterpreterUtils.cpp.o CMakeFiles/obj.clangInterpr
 │ │ eter.dir/Value.cpp.o CMakeFiles/obj.clangInterpreter.dir/Wasm.cpp.o
 │ │ $BUILD_PREFIX/opt/emsdk/upstream/emscripten/emranlib ../../../../lib/libclangInterpreter.a
 ```
 So
 1) this doesn't mention about any link libraries
 2) This gives us libclangInterpreter.a 
 
Now in the llvm build itself there might be some libXX.a that need libclangInterpreter.a (but then does libclangInterpreter.a convey the symbols of let's say clangAST to XX ?) 
 
 My use case is very much similar.
 
 1) I have A that need B that need to link against libclangInterpreter.a
 2) Now A can directly link to libclangInterpreter.a but I want to abstract A from it and only expose it through B
 3) Now as we know it is much easier to work with static libs for an emscripten/wasm context, so I was trying to create a libB.a 
 4) But then i am not sure as to how would LibB.a expose clangInterpreter to A
 
 So technically I am using a very similar script for B
 ```
 add_llvm_library(B
  // some source files

  # Specify LLVM components directly for linking
  LINK_COMPONENTS
  // Some components

  # Additional libraries from Clang and LLD
  LINK_LIBS
  clanginterpreter
)
```
Q1) Now should B be able to access symbols from clangInterpreter ? ( Maybe something like `emnm libB.a | grep "clang:Interpreter:Parse" for eg)
Q2) While building llvm against emscripten we would end up with so many static libraries (with one libraries having lot of others as Link_Libs) . So if that recipe is building smoothly why is A failing to access symbols out of C (or clangInterpreter) in my case ?
Really confused if there is some obvious flaw in my method.

Also @sbc100 really appreciate the detailed replies from your side !!

**Comment by sbc100** [comment]
In your example above `libB.a` would not contains `clanginterpreter` directly but `add_llvm_library` would encode a dependency such that anyone using libB would also end up implicitly linking `libclanginterpreter`.

If you inspect `libB.a` on disk you won't see any definitions from `libclanginterpreter`, but within C make it must be the case that users of B (via `add_llvm_library`) also end up being users of `libclanginterpreter` (indirectly).   i.e. the dependencies are encoded in the cmake build system, but not in the libraries themselves.

This is really a question about how `add_llvm_library` works I think.

**Comment by anutosh491** [comment]
> This is really a question about how add_llvm_library works I think.

Hmm, so you say things should work right out off the bat as in `add_llvm_library` does enough to indirectly add dependency on libclanginterpreter ?

Yeah I guess I was expecting that too (and it also works in the recipe as llvm obviously uses this call multiple times) but then when i build libA.a (that needs LibB.a) , I end up seeing lot of undefined symbols (some pasted below) and I know that all of these come out of libClangInterpreter.
```
warning: undefined symbol: _ZN5clang12HeaderSearch13AddSearchPathERKNS_15DirectoryLookupEb (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: _ZN5clang12LookupResult11deletePathsEPNS_12CXXBasePathsE (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: _ZN5clang12LookupResult11resolveKindEv (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: _ZN5clang12LookupResult9configureEv (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: _ZN5clang12TemplateNameC1EPNS_12TemplateDeclE (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: _ZN5clang13MangleContext10mangleNameENS_10GlobalDeclERN4llvm11raw_ostreamE (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: _ZN5clang13MangleContext20shouldMangleDeclNameEPKNS_9NamedDeclE (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: _ZN5clang16CompilerInstance20LoadRequestedPluginsEv (referenced by root reference (e.g. compiled C/C++ code))
```

This tells me that B is not exposing the symbols through C correctly.

**Comment by vgvassilev** [comment]
@sbc100, I think @anutosh491 is looking for a way to repackage a set of static libraries into a single giant static library that contains all llvm symbols he needs and then link against it.

**Comment by sbc100** [comment]
> @sbc100, I think @anutosh491 is looking for a way to repackage a set of static libraries into a single giant static library that contains all llvm symbols he needs and then link against it.

There is no simple way to do that that I know of, other than extracting all of the `.a` files (`emar x ....`) and then creating a new super-library from the resulting contents (`emar rcs ...`).   Again, this is not emscripten specific, its just how most toolchains work.

**Comment by anutosh491** [comment]
Thanks for the replies. Closing

**Comment by anutosh491** [comment]
Hi,
So I tried the approach as suggested above, this helps me get rid of any undefined symbols and now I atleast have all necessary symbols defined for my use case in the static library (clangCppInterOp or B as I mentioned above) but I still face an issue (so technically all required/essential symbols from C are available in B, but when I try to build A I get a `SIGSEGV` error)
```
[100%] Linking CXX executable xcpp.js
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/wasm-ld -o xcpp.wasm --whole-archive CMakeFiles/xcpp.dir/src/main_emscripten_kernel.cpp.o /Users/anutosh491/micromamba/envs/xeus-cpp-wasm-host/lib/libxeus-lite.a libxeus-cpp.a /Users/anutosh491/micromamba/envs/xeus-cpp-wasm-host/lib/libxeus.a /Users/anutosh491/micromamba/envs/xeus-cpp-wasm-host/lib/libclangCppInterOp.a /Users/anutosh491/micromamba/envs/xeus-cpp-wasm-host/lib/libpugixml.a -L/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/pic -lembind-rtti -lGL -lal -lhtml5 -lstubs -lc-asan -ldlmalloc -lcompiler_rt -lc++ -lc++abi -lsockets -lasan_rt -lasan_js -lubsan_rt -llsan_common_rt -lsanitizer_common_rt --no-whole-archive -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-cxx-exceptions -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --export-if-defined=setThrew --export-if-defined=malloc --export-if-defined=__errno_location --export-if-defined=htons --export-if-defined=ntohs --export-if-defined=memcpy --export-if-defined=free --export-if-defined=htonl --export-if-defined=_emscripten_timeout --export-if-defined=emscripten_builtin_malloc --export-if-defined=emscripten_builtin_free --export-if-defined=emscripten_builtin_memalign --export-if-defined=__cxa_is_pointer_type --export-if-defined=setTempRet0 --export-if-defined=__cxa_can_catch --export-if-defined=__cxa_increment_exception_refcount --export-if-defined=__cxa_decrement_exception_refcount --export-if-defined=__get_exception_message --export-if-defined=fileno --export-if-defined=__dl_seterr --export-if-defined=memcmp --export-if-defined=sleep --export-if-defined=__getTypeName /var/folders/m1/cdn74f917994jd99d_2cpf440000gn/T/tmpzt4d3u33libemscripten_js_symbols.so --import-undefined --import-memory --strip-debug --export-dynamic --export-if-defined=main --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=__main_argc_argv --export-if-defined=__wasm_apply_data_relocs --export-if-defined=fflush --export-if-defined=stackSave --export-if-defined=stackRestore --export-if-defined=stackAlloc --export-if-defined=__errno_location --export-if-defined=memalign --export-if-defined=emscripten_builtin_memalign --export-if-defined=emscripten_builtin_malloc --export-if-defined=emscripten_builtin_free --export-if-defined=_asan_c_load_1 --export-if-defined=_asan_c_load_1u --export-if-defined=_asan_c_load_2 --export-if-defined=_asan_c_load_2u --export-if-defined=_asan_c_load_4 --export-if-defined=_asan_c_load_4u --export-if-defined=_asan_c_load_f --export-if-defined=_asan_c_load_d --export-if-defined=_asan_c_store_1 --export-if-defined=_asan_c_store_1u --export-if-defined=_asan_c_store_2 --export-if-defined=_asan_c_store_2u --export-if-defined=_asan_c_store_4 --export-if-defined=_asan_c_store_4u --export-if-defined=_asan_c_store_f --export-if-defined=_asan_c_store_d --export-if-defined=__funcs_on_exit --export-if-defined=malloc --export-if-defined=free --export-if-defined=__cxa_is_pointer_type --export-if-defined=__cxa_can_catch --export-if-defined=__cxa_increment_exception_refcount --export-if-defined=__cxa_decrement_exception_refcount --export-if-defined=setThrew --export-if-defined=__cxa_free_exception --export-if-defined=__wasm_call_ctors --experimental-pic -pie -z stack-size=67108864 --initial-memory=575275008 --max-memory=2147483648 --no-entry
 #0 0x0000000102365e50 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x100125e50)
 #1 0x0000000102363e60 llvm::sys::RunSignalHandlers() (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x100123e60)
 #2 0x0000000102366a24 SignalHandler(int) (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x100126a24)
 #3 0x0000000195057584 (/usr/lib/system/libsystem_platform.dylib+0x180477584)
 #4 0x00000001026bc38c lld::wasm::ElemSection::writeBody() (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x10047c38c)
 #5 0x00000001026d4454 lld::wasm::SyntheticSection::finalizeContents() (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x100494454)
 #6 0x00000001026d0330 lld::wasm::(anonymous namespace)::Writer::run() (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x100490330)
 #7 0x00000001026c7174 lld::wasm::writeResult() (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x100487174)
 #8 0x00000001026960b8 lld::wasm::(anonymous namespace)::LinkerDriver::linkerMain(llvm::ArrayRef<char const*>) (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x1004560b8)
 #9 0x00000001026921ac lld::wasm::link(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, bool, bool) (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x1004521ac)
#10 0x0000000102377144 lld::unsafeLldMain(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, llvm::ArrayRef<lld::DriverDef>, bool) (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x100137144)
#11 0x00000001022c156c lld_main(int, char**, llvm::ToolContext const&) (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x10008156c)
#12 0x00000001022c1bf0 main (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x100081bf0)
#13 0x0000000194c9e0e0 
em++: error: '/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/wasm-ld -o xcpp.wasm --whole-archive CMakeFiles/xcpp.dir/src/main_emscripten_kernel.cpp.o /Users/anutosh491/micromamba/envs/xeus-cpp-wasm-host/lib/libxeus-lite.a libxeus-cpp.a /Users/anutosh491/micromamba/envs/xeus-cpp-wasm-host/lib/libxeus.a /Users/anutosh491/micromamba/envs/xeus-cpp-wasm-host/lib/libclangCppInterOp.a /Users/anutosh491/micromamba/envs/xeus-cpp-wasm-host/lib/libpugixml.a -L/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/pic -lembind-rtti -lGL -lal -lhtml5 -lstubs -lc-asan -ldlmalloc -lcompiler_rt -lc++ -lc++abi -lsockets -lasan_rt -lasan_js -lubsan_rt -llsan_common_rt -lsanitizer_common_rt --no-whole-archive -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-cxx-exceptions -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --export-if-defined=setThrew --export-if-defined=malloc --export-if-defined=__errno_location --export-if-defined=htons --export-if-defined=ntohs --export-if-defined=memcpy --export-if-defined=free --export-if-defined=htonl --export-if-defined=_emscripten_timeout --export-if-defined=emscripten_builtin_malloc --export-if-defined=emscripten_builtin_free --export-if-defined=emscripten_builtin_memalign --export-if-defined=__cxa_is_pointer_type --export-if-defined=setTempRet0 --export-if-defined=__cxa_can_catch --export-if-defined=__cxa_increment_exception_refcount --export-if-defined=__cxa_decrement_exception_refcount --export-if-defined=__get_exception_message --export-if-defined=fileno --export-if-defined=__dl_seterr --export-if-defined=memcmp --export-if-defined=sleep --export-if-defined=__getTypeName /var/folders/m1/cdn74f917994jd99d_2cpf440000gn/T/tmpzt4d3u33libemscripten_js_symbols.so --import-undefined --import-memory --strip-debug --export-dynamic --export-if-defined=main --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=__main_argc_argv --export-if-defined=__wasm_apply_data_relocs --export-if-defined=fflush --export-if-defined=stackSave --export-if-defined=stackRestore --export-if-defined=stackAlloc --export-if-defined=__errno_location --export-if-defined=memalign --export-if-defined=emscripten_builtin_memalign --export-if-defined=emscripten_builtin_malloc --export-if-defined=emscripten_builtin_free --export-if-defined=_asan_c_load_1 --export-if-defined=_asan_c_load_1u --export-if-defined=_asan_c_load_2 --export-if-defined=_asan_c_load_2u --export-if-defined=_asan_c_load_4 --export-if-defined=_asan_c_load_4u --export-if-defined=_asan_c_load_f --export-if-defined=_asan_c_load_d --export-if-defined=_asan_c_store_1 --export-if-defined=_asan_c_store_1u --export-if-defined=_asan_c_store_2 --export-if-defined=_asan_c_store_2u --export-if-defined=_asan_c_store_4 --export-if-defined=_asan_c_store_4u --export-if-defined=_asan_c_store_f --export-if-defined=_asan_c_store_d --export-if-defined=__funcs_on_exit --export-if-defined=malloc --export-if-defined=free --export-if-defined=__cxa_is_pointer_type --export-if-defined=__cxa_can_catch --export-if-defined=__cxa_increment_exception_refcount --export-if-defined=__cxa_decrement_exception_refcount --export-if-defined=setThrew --export-if-defined=__cxa_free_exception --export-if-defined=__wasm_call_ctors --experimental-pic -pie -z stack-size=67108864 --initial-memory=575275008 --max-memory=2147483648 --no-entry' failed (received SIGSEGV (-11))
make[2]: *** [xcpp.js] Error 1
make[1]: *** [CMakeFiles/xcpp.dir/all] Error 2
make: *** [all] Error 2
emmake: error: 'make install' failed (returned 2)
```

**Comment by anutosh491** [comment]
Hmmm this issue (https://github.com/emscripten-core/emscripten/issues/19774) talks about using the LTO flag and the issue being LTO related. Well in my case I don't see such a flag (not sure if it should even be used or not) but I have reported the error at llvm's upstream here https://github.com/llvm/llvm-project/issues/114651

**Comment by anutosh491** [comment]
Using 3.1.45-asserts I have some more info (not a lot though)
```
 #0 0x0000000106ccffac llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/lib/libLLVM.dylib+0x15bfac)
 #1 0x0000000106ccdfb4 llvm::sys::RunSignalHandlers() (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/lib/libLLVM.dylib+0x159fb4)
 #2 0x0000000106cd0620 SignalHandler(int) (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/lib/libLLVM.dylib+0x15c620)
 #3 0x0000000195057584 (/usr/lib/system/libsystem_platform.dylib+0x180477584)
 #4 0x0000000195026c20 (/usr/lib/system/libsystem_pthread.dylib+0x180446c20)
 #5 0x0000000194f33a30 (/usr/lib/system/libsystem_c.dylib+0x180353a30)
 #6 0x0000000194f32d20 (/usr/lib/system/libsystem_c.dylib+0x180352d20)
 #7 0x00000001024fb690 lld::wasm::FunctionSymbol::setFunctionIndex(unsigned int) (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x10031f690)
 #8 0x0000000102503cc4 lld::wasm::ElemSection::writeBody() (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x100327cc4)
 #9 0x000000010251ede8 lld::wasm::SyntheticSection::finalizeContents() (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x100342de8)
#10 0x000000010251aab4 lld::wasm::(anonymous namespace)::Writer::run() (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x10033eab4)
#11 0x0000000102512ad8 lld::wasm::writeResult() (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x100336ad8)
#12 0x00000001024d9a98 lld::wasm::(anonymous namespace)::LinkerDriver::linkerMain(llvm::ArrayRef<char const*>) (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x1002fda98)
#13 0x00000001024d62c0 lld::wasm::link(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, bool, bool) (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x1002fa2c0)
#14 0x00000001021def20 lld::unsafeLldMain(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, llvm::ArrayRef<lld::DriverDef>, bool) (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x100002f20)
#15 0x00000001021dda64 lld_main(int, char**, llvm::ToolContext const&) (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x100001a64)
#16 0x00000001021de0dc main (/Users/anutosh491/micromamba/envs/xeus-cpp-wasm-build/lib/python3.13/site-packages/emsdk/upstream/bin/lld+0x1000020dc)
#17 0x0000000194c9e0e0 
```

**Comment by sbc100** [comment]
That looks like a completely seperate issue so I think we should close this and continue the discussion on https://github.com/llvm/llvm-project/issues/114651
